File: sdk/ruby/lib/chain/authorization_grant.rb
require_relative './client_module'
require_relative './query'
require_relative './response_object'

module Chain
  class AuthorizationGrant < ResponseObject

    # @!attribute [r] guard_type
    # The type of credential that the guard matches against. Only "access_token"
    # and "x509" are allowed.
    # @return [String]
    attrib :guard_type

    # @!attribute [r] guard_data
    # A list of parameters that match specific credentials.
    # @return [Hash]
    attrib :guard_data

    # @!attribute [r] policy
    # @return [String]
    attrib :policy

    # @!attribute [r] protected
    # @return [Boolean]
    # Whether the grant can be deleted. Only used for internal purposes.
    attrib :protected

    # @!attribute [r] created_at
    # Timestamp of token creation.
    # @return [Time]
    attrib :created_at, rfc3339_time: true

    class ClientModule < Chain::ClientModule

      # Create an authorization grant, which provides the specified
      # credential with access to the given policy. Credentials are identified
      # using predicates called guards. Guards identify credentials by type
      # and by patterns specific to that type.
      #
      # @param [Hash] opts
      # @option opts [String] :guard_type Either "access_token" or "x509".
      # @option opts [Hash] :guard_data Parameters that describe a credential.
      #
      #   For guards of type "access_token", provide a Hash with a single key,
      #   "id", whose value is the unique ID of the access token.
      #
      #   For guards of type "x509", there should be a single top-level key,
      #   "subject", which maps to a hash of Subject attributes. Valid
      #   keys include:
      #     - "C" (Country, string or array of strings)
      #     - "O" (Organization, string or array of strings)
      #     - "OU" (Organizational Unit, string or array of strings)
      #     - "L" (Locality, string or array of strings)
      #     - "ST" (State or Province, string or array of strings)
      #     - "STREET" (Street Address, string or array of strings)
      #     - "POSTALCODE" (Postal Code, string or array of strings)
      #     - "SERIALNUMBER" (Serial Number, string)
      #     - "CN" (Common Name, string)
      #
      # @option opts [String] :policy One of the following:
      #
      #   - "client-readwrite": full access to the Client API, including
      #      accounts, assets, transactions, access tokens, MockHSM, etc.
      #   - "client-readonly": read-only access to the Client API.
      #   - "monitoring": read-only access to diagnostic components of the API,
      #      including fetching configuration info.
      #   - "crosscore": access to the cross-core API, including fetching blocks
      #      and submitting transactions, but not including block signing.
      #   - "crosscore-signblock": access to the cross-core API's block-signing
      #      API call.
      # @return [AuthorizationGrant]
      def create(opts)
        # Copy input and stringify keys
        opts = opts.reduce({}) do |memo, (k, v)|
          memo[k.to_s] = v
          memo
        end

        if opts['guard_type'].to_s == 'x509'
          opts['guard_data'] = self.class.sanitize_x509(opts['guard_data'])
        end

        AuthorizationGrant.new(client.conn.request('create-authorization-grant', opts))
      end

      # List all authorization grants. The sort order is not defined.
      # @return [Array<AuthorizationGrant>]
      def list_all
        client.conn.request('list-authorization-grants')['items'].map { |item| AuthorizationGrant.new(item) }
      end

      # Delete the specified authorization grant.
      # @param opts Identical to {#create}.
      # @return [void]
      def delete(opts)
        client.conn.request('delete-authorization-grant', opts)
        nil
      end

      SUBJECT_ATTRIBUTES = {
        'C' => {array: true},
        'O' => {array: true},
        'OU' => {array: true},
        'L' => {array: true},
        'ST' => {array: true},
        'STREET' => {array: true},
        'POSTALCODE' => {array: true},
        'SERIALNUMBER' => {array: false},
        'CN' => {array: false},
      }

      def self.sanitize_x509(guard_data)
        first_key = guard_data.keys.first
        if guard_data.size != 1 || first_key.to_s.downcase != 'subject'
          raise ArgumentError.new('Guard data must contain exactly one key, "subject"')
        end

        res = {}
        res[first_key] = guard_data.values.first.reduce({}) do |memo, (k, v)|
          attrib = SUBJECT_ATTRIBUTES[k.to_s.upcase]
          raise ArgumentError.new("Invalid subject attrib: #{k}") unless attrib

          if attrib[:array] && !v.is_a?(Array)
            memo[k] = [v]
          elsif !attrib[:array] && v.is_a?(Array)
            raise ArgumentError.new("Invalid array value for #{k}: #{v}")
          else
            memo[k] = v
          end

          memo
        end
        res
      end

    end

  end
end

File: sdk/java/src/main/java/com/chain/api/AuthorizationGrant.java
package com.chain.api;

import com.chain.exception.*;
import com.chain.http.*;

import java.util.*;

import com.google.gson.annotations.SerializedName;

/**
 * An authorization grant provides access to resources in the Chain Core API.
 * It maps guards (predicates that match against credentials provided by
 * clients) to policies (lists of resources).
 * <p>
 * There are two types of guards: {@link AccessTokenGuard}, which matches against
 * access tokens by ID, and {@link X509Guard}, which matches against X.509 client
 * certificates that match a set of attributes.
 * <p>
 * Currently, there are four policies exposed through the API:
 * <p><ul>
 * <li>client-readwrite: full access to the Client API, including accounts,
 *   assets, transactions, access tokens, authorization grants, etc.
 * <li>client-readonly: read-only access to the Client API. API calls that modify
 *   data in Chain Core, such as account creation, are not permitted.
 * <li>monitoring: read-only access to monitoring endpoints, such as fetching
 *   the Chain Core configuration.
 * <li>crosscore: access to the cross-core API, including fetching blocks and
 *   submitting transactions, but not including block signing.
 * <li>crosscore-signblock: access to the cross-core API's block-signing API call.
 * </ul>
 */
public class AuthorizationGrant {
  @SerializedName("guard_type")
  public String guardType;

  @SerializedName("guard_data")
  public Map<String, Object> guardData;

  public String policy;

  @SerializedName("protected")
  public boolean isProtected;

  @SerializedName("created_at")
  public Date createdAt;

  /**
   * A guard that will provide access for a specific access token, identified
   * by its unique ID.
   */
  public static class AccessTokenGuard {
    public String id;

    /**
     * Specifies the ID of the token that the guard will match against.
     * @param id an access token ID (just the ID, not the full token value)
     * @return updated AccessTokenGuard object
     */
    public AccessTokenGuard setId(String id) {
      this.id = id;
      return this;
    }
  }

  /**
   * A guard that will provide access for X.509 certificates whose Subject
   * attribute matches a specified list of sub-attributes, such as CN or OU.
   * If a certificate's Subject contains all of the sub-attribute values
   * specified in the guard, the guard will produce a positive match.
   */
  public static class X509Guard {
    public static class Subject {
      @SerializedName("C")
      List<String> country;

      @SerializedName("O")
      List<String> organization;

      @SerializedName("OU")
      List<String> organizationalUnit;

      @SerializedName("L")
      List<String> locality;

      @SerializedName("ST")
      List<String> stateOrProvince;

      @SerializedName("STREET")
      List<String> streetAddress;

      @SerializedName("POSTALCODE")
      List<String> postalCode;

      @SerializedName("SERIALNUMBER")
      String serialNumber;

      @SerializedName("CN")
      String commonName;
    }

    private Subject subject;

    public X509Guard() {
      subject = new Subject();
    }

    /**
     * Specifies a value for the Country attribute. Multiple values can be specified.
     * @param s value for Country attribute
     * @return updated guard object
     */
    public X509Guard addCountry(String s) {
      if (subject.country == null) {
        subject.country = new ArrayList<>();
      }
      subject.country.add(s);
      return this;
    }

    /**
     * Specifies a value for the Organization attribute. Multiple values can be specified.
     * @param s value for Organization attribute
     * @return updated guard object
     */
    public X509Guard addOrganization(String s) {
      if (subject.organization == null) {
        subject.organization = new ArrayList<>();
      }
      subject.organization.add(s);
      return this;
    }

    /**
     * Specifies a value for the Organizational Unit attribute. Multiple values can be specified.
     * @param s value for Organizational Unit attribute
     * @return updated guard object
     */
    public X509Guard addOrganizationalUnit(String s) {
      if (subject.organizationalUnit == null) {
        subject.organizationalUnit = new ArrayList<>();
      }
      subject.organizationalUnit.add(s);
      return this;
    }

    /**
     * Specifies a value for the Locality attribute. Multiple values can be specified.
     * @param s value for Locality attribute
     * @return updated guard object
     */
    public X509Guard addLocality(String s) {
      if (subject.locality == null) {
        subject.locality = new ArrayList<>();
      }
      subject.locality.add(s);
      return this;
    }

    /**
     * Specifies a value for the State/Province attribute. Multiple values can be specified.
     * @param s value for State/Province attribute
     * @return updated guard object
     */
    public X509Guard addStateOrProvince(String s) {
      if (subject.stateOrProvince == null) {
        subject.stateOrProvince = new ArrayList<>();
      }
      subject.stateOrProvince.add(s);
      return this;
    }

    /**
     * Specifies a value for the Street Address attribute. Multiple values can be specified.
     * @param s value for Street Address attribute
     * @return updated guard object
     */
    public X509Guard addStreetAddress(String s) {
      if (subject.streetAddress == null) {
        subject.streetAddress = new ArrayList<>();
      }
      subject.streetAddress.add(s);
      return this;
    }

    /**
     * Specifies a value for the Postal Code attribute. Multiple values can be specified.
     * @param s value for Postal Code attribute
     * @return updated guard object
     */
    public X509Guard addPostalCode(String s) {
      if (subject.postalCode == null) {
        subject.postalCode = new ArrayList<>();
      }
      subject.postalCode.add(s);
      return this;
    }

    /**
     * Specifies a value for the Serial Number attribute.
     * @param s value for Serial Number attribute
     * @return updated guard object
     */
    public X509Guard setSerialNumber(String s) {
      subject.serialNumber = s;
      return this;
    }

    /**
     * Specifies a value for the Common Name attribute.
     * @param s value for Common Name attribute
     * @return updated guard object
     */
    public X509Guard setCommonName(String s) {
      subject.commonName = s;
      return this;
    }
  }

  /**
   * A base class for RPC builders that specify a grant, i.e. a
   * guard-policy tuple.
   * @param <T> always use the child class that extends BaseBuilder
   */
  public static class BaseBuilder<T> {
    @SerializedName("guard_type")
    private String guardType;

    @SerializedName("guard_data")
    private Object guardData;

    private String policy;

    /**
     * Specifies a guard that will match against a specific access token.
     * @param g a guard that matches against an access token
     * @return updated builder object
     */
    public T setGuard(AccessTokenGuard g) {
      this.guardType = "access_token";
      this.guardData = g;
      return (T) this;
    }

    /**
     * Specifies a guard that will match against X.509 certificates.
     * @param g a guard that matches against X.509 certificates
     * @return updated builder object
     */
    public T setGuard(X509Guard g) {
      guardType = "x509";
      guardData = g;
      return (T) this;
    }

    /**
     * Sets the policy to grant to credentials that match the guard.
     * @param policy One of "client-readwrite", "client-readonly", "monitoring", or "network"
     * @return updated builder object
     */
    public T setPolicy(String policy) {
      this.policy = policy;
      return (T) this;
    }
  }

  /**
   * Sets up a grant creation API call.
   */
  public static class Builder extends BaseBuilder<Builder> {
    /**
     * Creates a new grant with the parameters in this Builder instance.
     * @param client the client object providing connectivity to the Chain Core instance
     * @throws ChainException
     */
    public void create(Client client) throws ChainException {
      client.request("create-authorization-grant", this, SuccessMessage.class);
    }
  }

  /**
   * Sets up a grant deletion API call.
   */
  public static class DeletionBuilder extends BaseBuilder<DeletionBuilder> {
    /**
     * Deletes a new grant with the parameters in this Builder instance.
     * @param client the client object providing connectivity to the Chain Core instance
     * @throws ChainException
     */
    public void delete(Client client) throws ChainException {
      client.request("delete-authorization-grant", this, SuccessMessage.class);
    }
  }

  /**
   * Retrieves a list of all authorization grants in Chain Core.
   * @param client the client object providing connectivity to the Chain Core instance
   * @return a list of authorization grants
   * @throws ChainException
   */
  public static List<AuthorizationGrant> listAll(Client client) throws ChainException {
    ListResponse resp = client.request("list-authorization-grants", null, ListResponse.class);
    return resp.items;
  }

  private class ListResponse {
    public java.util.List<AuthorizationGrant> items;
  }
}
